snapshots/devmapper/snapshotter.go

```diff
@@ -398,26 +410,87 @@ func (s *Snapshotter) createSnapshot(ctx context.Context, kind snapshots.Kind, k
 	return mounts, nil
 }
 
-// mkfs creates ext4 filesystem on the given devmapper device
-func mkfs(ctx context.Context, path string) error {
-	args := []string{
-		"-E",
-		// We don't want any zeroing in advance when running mkfs on thin devices (see "man mkfs.ext4")
-		"nodiscard,lazy_itable_init=0,lazy_journal_init=0",
-		path,
+// mkfs creates filesystem on the given devmapper device based on type
+// specified in config. If no file system type is specified in config, ext4
+// is used as the default file system type
+func mkfs(ctx context.Context, fs fsType, path string) error {
+	mkfsCommand := ""
+	var args []string
+
+	switch fs {
+	case fsTypeExt4:
+		mkfsCommand = "mkfs.ext4"
+		args = []string{
+			"-E",
+			// We don't want any zeroing in advance when running mkfs on thin devices (see "man mkfs.ext4")
+			"nodiscard,lazy_itable_init=0,lazy_journal_init=0",
+			path,
+		}
+	case fsTypeXfs:
+		mkfsCommand = "mkfs.xfs"
+		args = []string{
+			path,
+		}
+	default:
+		return errors.New("file system not supported")
 	}
 
-	log.G(ctx).Debugf("mkfs.ext4 %s", strings.Join(args, " "))
-	b, err := exec.Command("mkfs.ext4", args...).CombinedOutput()
+	log.G(ctx).Debugf("%s %s", mkfsCommand, strings.Join(args, " "))
+	b, err := exec.Command(mkfsCommand, args...).CombinedOutput()
 	out := string(b)
 	if err != nil {
-		return errors.Wrapf(err, "mkfs.ext4 couldn't initialize %q: %s", path, out)
+		return errors.Wrapf(err, "%s couldn't initialize %q: %s", mkfsCommand, path, out)
 	}
 
 	log.G(ctx).Debugf("mkfs:\n%s", out)
 	return nil
 }
 
+func (s *Snapshotter) getFsType(device string) string {
+	fsSigs := []FsSigData{
+		// ext4 filesystem always has the bytes 0x53 0xEF at positions 1080–1081 (0x438)
+		{"ext4", "\123\357", 0x438},
+		// XFS magic number is in bytes 0 to 3 and has the byte sequence "XFSB" at offset 0
+		{"xfs", "XFSB", 0},
+	}
+
+	maxReadLen := uint64(0)
+	for _, f := range fsSigs {
+		l := f.offset + uint64(len(f.magicStr))
+		if l > maxReadLen {
+			maxReadLen = l
+		}
+	}
+
+	file, err := os.Open(device)
+	if err != nil {
+		fmt.Printf("error opening device %s", device)
+		return ""
+	}
+
+	buf := make([]byte, maxReadLen)
+	l, err := file.Read(buf)
```

Comment:
The reason we have to use the max of all the offsets ( for each FS type supported ) is because we dont know the type before hand, and hence copy the maximum len that is guaranteed to have signature for the supported FS types. In this case, i think copying it once and working on the buffer would be faster than multiple seek into the file, as seeks are inherently slow operations.

Reply:
I see. I didn't know seek is slow